test/fixtures/azure-fixture/src/main/java/fixture/azure/MockAzureBlobStore.java

```diff
@@ -0,0 +1,483 @@
+                        logger.debug("Mismatch on release - submitted leaseId: {}, active leaseId: {}", requestLeaseId, this.leaseId);
+                        throw new ConflictException(
+                            "LeaseIdMismatchWithLeaseOperation",
+                            "The lease ID specified did not match the lease ID for the blob/container."
+                        );
+                    }
+                    state = State.Available;
+                    this.leaseId = null;
+                    this.expireTimeMillisSinceEpoch = Long.MAX_VALUE;
+                }
+            }
+        }
+
+        public synchronized void assertCanWrite(@Nullable String requestLeaseId) {
+            expireIfDue();
+            switch (state) {
+                case Available, Expired -> {
+                    if (requestLeaseId != null) {
+                        throw new PreconditionFailedException(
+                            "LeaseLost",
+                            "A lease ID was specified, but the lease for the blob/container has expired."
+                        );
+                    }
+                }
+                case Leased -> {
+                    if (requestLeaseId == null) {
+                        throw new PreconditionFailedException(
+                            "LeaseIdMissing",
+                            "There is currently a lease on the blob/container and no lease ID was specified in the request."
+                        );
+                    }
+                    if (leaseId.equals(requestLeaseId) == false) {
+                        throw new ConflictException(
+                            "LeaseIdMismatchWithBlobOperation",
+                            "The lease ID specified did not match the lease ID for the blob."
+                        );
+                    }
+                }
+            }
+        }
+
+        public synchronized void assertCanRead(@Nullable String requestLeaseId) {
+            expireIfDue();
+            switch (state) {
+                case Available, Expired -> {
+                    if (requestLeaseId != null) {
+                        throw new PreconditionFailedException(
+                            "LeaseLost",
+                            "A lease ID was specified, but the lease for the blob/container has expired."
+                        );
+                    }
+                }
+                case Leased -> {
+                    if (requestLeaseId != null && requestLeaseId.equals(leaseId) == false) {
+                        throw new ConflictException(
+                            "LeaseIdMismatchWithBlobOperation",
+                            "The lease ID specified did not match the lease ID for the blob."
+                        );
+                    }
+                }
+            }
+        }
+
+        private void expireIfDue() {
+            if (state == State.Leased && System.currentTimeMillis() >= expireTimeMillisSinceEpoch) {
```

Comment:
I think for testing purposes we probably don't want to have time-based expiry. Instead, we should generally not expire leases, but then maybe have some tests where leases expire randomly (only on contended register operations) to ensure we do the right thing in those cases.

Reply:
I added support for failing the lease in 2b2b87c88e021ec0b441e8b659b4ba9d119d19b5

I tried configuring the contended tests to fail the lease occasionally on contention and it showed that we might erroneously fail the compare-and-swap when the lease is expired between the update and the release. I fixed that in 26223a179960426a63c815022988cb4649efb20c

I wanted to add [this link](https://learn.microsoft.com/en-us/rest/api/storageservices/lease-blob?outcomes-of-lease-operations-on-blobs-by-lease-state) to `bestEffortRelease` but checkstyle had other ideas